compileopts: re-enable support for GOARCH=wasm

aykevl opened the pull request:
Support for GOARCH=wasm was broken entirely. With this PR, it should work again for supported values of GOOS.

This fixes the following error for example:

    $ make tinygo-test-wasip1
    GOOS=wasip1 GOARCH=wasm /home/ayke/bin/tinygo test cmp compress/lzw compress/zlib [...etc]
    cannot resolve packages: GOARCH=wasm but GOOS is unset. Please set GOOS to wasm, wasip1, or wasip2.
    make: *** [GNUmakefile:489: tinygo-test-wasip1] Error 1

aykevl committed:
compileopts: enable support for GOARCH=wasm in `tinygo test`

Support for GOARCH=wasm was only available for `tinygo build`, not for
any of the other subcommands (`test`, `run`, etc). With this PR, these
subcommands should work again for supported values of GOOS.

This fixes the following error for example:

    $ make tinygo-test-wasip1
    GOOS=wasip1 GOARCH=wasm /home/ayke/bin/tinygo test cmp compress/lzw compress/zlib [...etc]
    cannot resolve packages: GOARCH=wasm but GOOS is unset. Please set GOOS to wasm, wasip1, or wasip2.
    make: *** [GNUmakefile:489: tinygo-test-wasip1] Error 1

deadprogram commented on compileopts/target.go:
```diff
@@ -179,7 +179,15 @@ func (spec *TargetSpec) resolveInherits() error {
 func LoadTarget(options *Options) (*TargetSpec, error) {
 	if options.Target == "" {
-		return defaultTarget(options)
+		if options.GOARCH == "wasm" && options.GOOS == "js" {
```
I probably would have written it using a switch:

```go
		switch {
		case options.GOARCH == "wasm" && options.GOOS == "js":
			options.Target = "wasm"
		case options.GOARCH == "wasm" && options.GOOS == "wasip1":
			options.Target = "wasip1"
		case options.GOARCH == "wasm" && options.GOOS == "wasip2":
			options.Target = "wasip2"
		default:
			return defaultTarget(options)
		}
```

aykevl replied:
Rewritten this as a switch now, in the updated version of this PR.

ydnar commented on compileopts/target.go:
```diff
@@ -179,7 +179,15 @@ func (spec *TargetSpec) resolveInherits() error {
+			options.Target = "wasm"
+		} else if options.GOARCH == "wasm" && options.GOOS == "wasip1" {
+			options.Target = "wasip1"
+		} else if options.GOARCH == "wasm" && options.GOOS == "wasip2" {
```
Will setting `GOOS=wasip2` even work? Will Go choke on the unknown `GOOS`?

aykevl replied:
Yes, it will work. I've tested it. GOOS is rewritten to what is set in targets/wasip2.json before `go` commands are invoked.

aykevl commented:
Reworked the PR. I realize now why `tinygo test` wasn't working: the special handling for GOOS=wasm was only done for `tinygo build` (not for any of the other subcommands). I've removed the old handling and moved it to LoadTarget which is a more appropriate place.

deadprogram commented:
Thanks for the fix @aykevl now merging.
